### PR TITLE
Re-implement DataFrame unboxing

### PR DESCRIPTION
**Pull request by akharche**

### CODE REVIEW

File: sdc/hiframes/pd_dataframe_ext.py [hunk 1]
```diff
@@ -59,6 +59,32 @@ class ColumnLoc(NamedTuple):
     col_id: int
 
 
+def get_columns_loc(col_types, col_names):
```
Comment (densmirn) [inline sdc/hiframes/pd_dataframe_ext.py:62]: Let's rename the function. Maybe `get_structure_maps`?

File: sdc/hiframes/boxing.py [hunk 2]
```diff
@@ -100,87 +79,46 @@ def typeof_pd_str_series(val, c):
...
-
-    def wrapper(args):
-        return my_intrinsic(args)
-
-    args = [column_tup]
     col_list_type = types.List(string_type)
-    resty = col_list_type
-    uni_tup = types.UniTuple(string_type, n_cols)
-    sig = typing.signature(resty, uni_tup)
+    _, inst = listobj.ListInstance.allocate_ex(c.context, c.builder, col_list_type, n_cols)
```
Comment (densmirn) [inline sdc/hiframes/boxing.py:94]: Maybe handle allocating of the list something like that:
```python
errorptr = cgutils.alloca_once_value(c.builder, cgutils.false_bit)
ok, inst = listobj.ListInstance.allocate_ex(c.context, c.builder, col_list_type, n_cols)
with c.builder.if_else(ok, likely=True) as (if_ok, if_not_ok):
    with if_ok:
        # code where inst is used
    with if_not_ok:
        c.builder.store(cgutils.true_bit, errorptr)

# If an error occurred, drop the whole native list
with c.builder.if_then(c.builder.load(errorptr)):
    c.context.nrt.decref(c.builder, col_list_type, inst.value)
```

File: sdc/hiframes/pd_dataframe_ext.py [hunk 3]
```diff
@@ -59,6 +59,32 @@ class ColumnLoc(NamedTuple):
...
+    # Define df structure, map column name to column position ex. {'A': (0,0), 'B': (1,0), 'C': (0,1)}
```
[the remark this answers is not in the document]
Comment (akharche) [reply sdc/hiframes/pd_dataframe_ext.py]: done

File: sdc/hiframes/pd_dataframe_ext.py [hunk 4]
```diff
@@ -59,6 +59,32 @@ class ColumnLoc(NamedTuple):
...
+        if col_typ not in data_typs_map:
+            data_typs_map[col_typ] = (type_id, [col_id])
+            # The first column in each type always has 0 index
+            column_loc[col_name] = ColumnLoc(type_id, 0)
+            types_order.append(col_typ)
+            type_id += 1
+        else:
+            # Get index of column in list of types
+            type_idx, col_indices = data_typs_map[col_typ]
+            col_idx_list = len(col_indices)
```
[the remark this answers is not in the document]
Comment (akharche) [reply sdc/hiframes/pd_dataframe_ext.py]: done

File: sdc/hiframes/pd_dataframe_ext.py (hunk 1 above)
Comment (akharche) [reply sdc/hiframes/pd_dataframe_ext.py]: done

File: sdc/hiframes/boxing.py (hunk 2 above)
Comment (akharche) [reply sdc/hiframes/boxing.py]: done

File: sdc/tests/test_dataframe.py [hunk 5]
```diff
@@ -184,7 +182,7 @@ def test_impl(df):
         df = pd.DataFrame({'A': ['a', 'b', 'c']})
         self.assertEqual(hpat_func(df), test_impl(df))
 
-    @dfRefactoringNotImplemented
+    @dfRefactoringNotImplemented  # Need boxing
```
Comment (densmirn) [inline sdc/tests/test_dataframe.py:185]: Let's remove the decorator.

File: sdc/tests/test_dataframe.py [hunk 6]
```diff
@@ -1493,7 +1481,7 @@ def test_impl(n):
             with self.subTest(n=n):
                 pd.testing.assert_frame_equal(sdc_impl(n), test_impl(n))
 
-    @dfRefactoringNotImplemented  # required re-implementing DataFrame unboxing
+    @dfRefactoringNotImplemented  # required re-implementing DataFrame boxing
```
Comment (densmirn) [inline sdc/tests/test_dataframe.py:1484]: Let's remove the decorator too.

**Pull request merged**